amdsharktuner/amdsharktuner/constraint_generator.py

```diff
@@ -7,14 +7,136 @@
 
 from . import common, dispatch_constraints, dispatch_parser
 
 
+class Z3Vals(Protocol):
+    """Marker base class for values extracted from Z3 model."""
+
+
+class Z3Vars(Protocol):
+    """Marker base class for Z3 variables."""
+
+    @property
+    def all_vars(self) -> list[z3.ExprRef]:
+        """All Z3 variables that participate in this problem."""
+        pass
+
+    def eval(self, model: z3.ModelRef) -> Z3Vals:
+        """Convert Z3 variables in the given model to values."""
+        pass
+
+
+@dataclass
+class ContractionZ3Vals(Z3Vals):
+    m_vals: list[int]
+    n_vals: list[int]
+    k_vals: list[int]
+    subgroup_m_vals: list[int]
+    subgroup_n_vals: list[int]
+
+    subgroup_size: int
+    intrinsic_mn: int
+    intrinsic_k: int
+    wg_x: int
+    wg_y: int
+    wg_z: int
+    sg_m_cnt: int
+    sg_n_cnt: int
+
+
+@dataclass
+class ContractionZ3Vars(Z3Vars):
+    m_vars: list[z3.ExprRef]
+    n_vars: list[z3.ExprRef]
+    k_vars: list[z3.ExprRef]
+    subgroup_m_vars: list[z3.ExprRef]
+    subgroup_n_vars: list[z3.ExprRef]
+
+    subgroup_size: z3.ExprRef
+    intrinsic_mn: z3.ExprRef
+    intrinsic_k: z3.ExprRef
+    wg_x: z3.ExprRef
+    wg_y: z3.ExprRef
+    wg_z: z3.ExprRef
+    sg_m_cnt: z3.ExprRef
+    sg_n_cnt: z3.ExprRef
+
+    @classmethod
+    def from_sizes(cls, matmul_size: common.ContractionSizes) -> "ContractionZ3Vars":
+        M, N, K = matmul_size.M, matmul_size.N, matmul_size.K
+
+        m_vars = [z3.Int(f"m{i}") for i in range(len(M))]
+        n_vars = [z3.Int(f"n{i}") for i in range(len(N))]
+        k_vars = [z3.Int(f"k{i}") for i in range(len(K))]
+        subgroup_m_vars = [z3.Int(f"subgroup_m{i}") for i in range(len(M))]
+        subgroup_n_vars = [z3.Int(f"subgroup_n{i}") for i in range(len(N))]
+
+        subgroup_size = z3.Int("subgroup_size")
+        intrinsic_mn = z3.Int("intrinsic_mn")
+        intrinsic_k = z3.Int("intrinsic_k")
+        wg_x, wg_y, wg_z = z3.Int("wg_x"), z3.Int("wg_y"), z3.Int("wg_z")
+        sg_m_cnt = z3.Int("sg_m_cnt")
+        sg_n_cnt = z3.Int("sg_n_cnt")
+
+        return cls(
+            m_vars=m_vars,
+            n_vars=n_vars,
+            k_vars=k_vars,
+            subgroup_m_vars=subgroup_m_vars,
+            subgroup_n_vars=subgroup_n_vars,
+            subgroup_size=subgroup_size,
+            intrinsic_mn=intrinsic_mn,
+            intrinsic_k=intrinsic_k,
+            wg_x=wg_x,
+            wg_y=wg_y,
+            wg_z=wg_z,
+            sg_m_cnt=sg_m_cnt,
+            sg_n_cnt=sg_n_cnt,
+        )
+
+    @property
+    def all_vars(self) -> list[z3.ExprRef]:
+        vars_list: list[z3.ExprRef] = []
+        for f in fields(self):
+            attr = getattr(self, f.name)
+            if isinstance(attr, list):
+                vars_list.extend(attr)
+            else:
+                vars_list.append(attr)
+        return vars_list
+
+    def eval(self, model: z3.ModelRef) -> ContractionZ3Vals:
+        get = lambda v: model[v].as_long()
```

Comment:
There's also `m.eval(x)` in the z3 API -- would using that help? https://theory.stanford.edu/~nikolaj/programmingz3.html#sec-models